### ISSUE TITLE
Connection retry doesn't always work after RabbitMQ is killed and restarted

### ISSUE THREAD

**Issue by jliu-tera** [guidesmiths/rascal#118]
Or more accurately most times it doesn't work.

Below is the broker config:

```
module.exports = {
  vhosts: {
    [config.RABBITMQ_VHOST]: {
      connection: {
        url: `${config.RABBITMQ_CONNECTION_STRING}`
      },
      exchanges: {
        [config.RABBITMQ_EXCHANGE_ENDPOINT_OUT]: {
          type: "topic",
          assert: false,
          check: true
        }
      },
      publications: {
        connectorPub: {
          exchange: `${config.RABBITMQ_EXCHANGE_ENDPOINT_OUT}`,
          routingKey: `${config.RABBITMQ_PUBLISH_ROUTING_KEY}`
        }
      }
    }
  }
};
```
I have 3 issues:
1. Rascal doesn't retry to connection if RabbitMQ has not been started (Rascal is started earlier than RabbitMQ). We mitigate this issue by exiting and let Kubernetes to restart the pods. Eventually Rascal will connect after RabbitMQ is started.
2. After Rascal is connected to RabbitMQ, kill and restart RabbitMQ. Sometimes it will retry to connect, but most times it won't. Connection reset error is provided though, but no retry.
{"message":"Broker error: Error: write ECONNRESET; vhost: /; connectionUrl: amqp://guest:***@localhost:31072","level":"error"}
3. If RabbitMQ is killed twice, Rascal will just give up. No retry has been observed.

I'm using the default retry configuration and I hope it should work as assumed. Please let me know whether this is a configuration issue or a bug.

Thanks.

**Comment by willthrom** [comment]
Hi,

@jliu-tera  first issue, Rascal generates an exception if the cluster is down. Rascal broker is only initialised if can connect to the Cluster first, if not an exception is throw and you need to do the reconnect manually. (We manage it with Rancher killing the App if RabbitMQ was not available

I will create a separate ticket as for Cluster configuration (multi-node other issues are happening).

**Comment by cressie176** [comment]
Hi @jliu-tera, as @willthrom points out, 1. is intended behaviour. RabbitMQ *should* be more highly available than your applications (one of the reasons for using a message broker in the first place), therefore if it is unavailable when your application first starts, it's mostly likely to be an unrecoverable configuration error. Based on the principle of fail fast and loud, Rascal throws an error, and wont automatically reconnect.

I don't see the behaviour you describe in 2 & 3 so I suspect they are configuration issues. Try running with the `DEBUG` environment variable set to `rascal:*` to see what it is doing under the hood.

**Comment by jliu-tera** [comment]
Thanks for the explanation. The way how Rascal handles scenario 1 is an acceptable solution. 

I will try the DEBUG to get more log for scenarios 2 and 3.

By the way, the version I'm using:  "rascal": "^10.2.1" if that matters.

**Comment by jliu-tera** [comment]
Hi Below is the debug log from killing RabbitMQ. As you can see there was no re-tries after connection was reset. 

Again I'm using default configurations for re-try. What else could be done to make it configured right?

```
  rascal:config:configure Configuring vhost: / +0ms
  rascal:config:configure Configuring exchange: endpoint.out +4ms
  rascal:config:configure Configuring exchange: ditto.out +1ms
  rascal:config:configure Configuring exchange:  +0ms
  rascal:config:configure Configuring queue: ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +0ms
  rascal:config:configure Configuring binding: bind-to-fanout +1ms
  rascal:config:configure Configuring publication: connectorPub +0ms
  rascal:config:configure Configuring subscription: connectorSub +1ms
  rascal:config:configure Configuring publication: connectorPub +0ms
  rascal:config:configure Configuring publication: /endpoint.out +0ms
  rascal:config:configure Configuring publication: /ditto.out +0ms
  rascal:config:configure Configuring publication: / +0ms
  rascal:config:configure Configuring publication: /ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +0ms
  rascal:config:configure Configuring subscription: connectorSub +1ms
  rascal:config:configure Configuring subscription: /ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +0ms
  rascal:config:configure Configuring counter: stub +0ms
  rascal:Broker Initialising broker +0ms
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072 +0ms
  rascal:tasks:createConnection Obtained connection: 4921b666-14d2-4031-bf14-72aafdb0434e +20ms
  rascal:tasks:createChannel Creating channel +0ms
  rascal:tasks:checkExchanges Checking exchange: endpoint.out +0ms
  rascal:tasks:checkExchanges Checking exchange: ditto.out +2ms
  rascal:tasks:assertQueues Asserting queue: ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +0ms
  rascal:tasks:applyBindings Binding queue: ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 to exchange: ditto.out with binding key: command +0ms
  rascal:tasks:closeChannel Closing channel +0ms
  rascal:Vhost Creating regular channel pool { autostart: false, max: 5, min: 1, evictionRunIntervalMillis: 10000, idleTimeoutMillis: 60000, testOnBorrow: true } +35ms
  rascal:Vhost Creating confirm channel pool { autostart: false, max: 5, min: 1, evictionRunIntervalMillis: 10000, idleTimeoutMillis: 60000, testOnBorrow: true } +3ms
  rascal:Vhost vhost: / was initialised with connection: 4921b666-14d2-4031-bf14-72aafdb0434e +1ms
  rascal:Publication Initialising publication: connectorPub +0ms
  rascal:Publication Initialising publication: /endpoint.out +0ms
  rascal:Publication Initialising publication: /ditto.out +0ms
  rascal:Publication Initialising publication: / +0ms
  rascal:Publication Initialising publication: /ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +0ms
  rascal:Subscription Initialising subscription: connectorSub +0ms
  rascal:Subscription Initialising subscription: /ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +0ms
  rascal:Subscription Subscribing to queue: ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +3ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +5ms
  rascal:Vhost Created confirm channel: a5a5deae-7d14-429d-9a15-9e4c42856b2a from connection: 4921b666-14d2-4031-bf14-72aafdb0434e +6ms
  rascal:SubscriberSession Opening subscriber session: amq.ctag-LFVNc3-m9rbiu_0Bt0aREw on channel: a5a5deae-7d14-429d-9a15-9e4c42856b2a +0ms
{"message":"Subscribed to Ditto messages","level":"info"}
  rascal:Vhost Handling connection error: write ECONNRESET initially from connection: 4921b666-14d2-4031-bf14-72aafdb0434e, amqp://guest:***@localhost:31072 +26s
{"message":"Broker error: Error: write ECONNRESET; vhost: /; connectionUrl: amqp://guest:***@localhost:31072","level":"error"}
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072 +26s
  rascal:Subscription Handling channel error: write ECONNRESET from connectorSub using channel: a5a5deae-7d14-429d-9a15-9e4c42856b2a +26s
{"message":"Subscriber error: Error: write ECONNRESET","level":"error"}
  rascal:SubscriberSession Removing channel: a5a5deae-7d14-429d-9a15-9e4c42856b2a from session +26s
  rascal:Subscription Subscribing to queue: ec-in-queue-a381ff92-f12c-4f91-938c-fcb444dd6f29 +2ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +3ms
```

**Comment by cressie176** [comment]
Thanks for the debug @jliu-tera. I've tried to reproduce locally using the simple example (modified so it doesn't publish messages in order to keep the output simpler), and by executing `docker kill` against a local RabbitMQ container. Everything works as expected, even if I try multiple times...

```
$ DEBUG='rascal:*' node index.js 
  rascal:config:configure Configuring vhost: / +0ms
  rascal:config:configure Configuring exchange: demo_ex +5ms
  rascal:config:configure Configuring exchange:  +0ms
  rascal:config:configure Configuring queue: demo_q +1ms
  rascal:config:configure Configuring binding: demo_ex[a.b.c] -> demo_q +1ms
  rascal:config:configure Configuring publication: demo_pub +0ms
  rascal:config:configure Configuring subscription: demo_sub +1ms
  rascal:config:configure Configuring publication: demo_pub +0ms
  rascal:config:configure Configuring publication: /demo_ex +0ms
  rascal:config:configure Configuring publication: / +0ms
  rascal:config:configure Configuring publication: /demo_q +1ms
  rascal:config:configure Configuring subscription: demo_sub +0ms
  rascal:config:configure Configuring subscription: /demo_q +0ms
  rascal:config:configure Configuring counter: stub +0ms
  rascal:Broker Initialising broker +0ms
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +0ms
  rascal:tasks:createConnection Obtained connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e +35ms
  rascal:tasks:createChannel Creating channel +0ms
  rascal:tasks:assertExchanges Asserting exchange: demo_ex +0ms
  rascal:tasks:assertQueues Asserting queue: demo_q +0ms
  rascal:tasks:applyBindings Binding queue: demo_q to exchange: demo_ex with binding key: a.b.c +0ms
  rascal:tasks:closeChannel Closing channel +0ms
  rascal:Vhost Creating regular channel pool { autostart: false, max: 5, min: 1, evictionRunIntervalMillis: 10000, idleTimeoutMillis: 60000, testOnBorrow: true } +66ms
  rascal:Vhost Creating confirm channel pool { autostart: true, max: 5, min: 1, evictionRunIntervalMillis: 10000, idleTimeoutMillis: 60000, testOnBorrow: true } +3ms
  rascal:Vhost vhost: / was initialised with connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e +1ms
  rascal:Publication Initialising publication: demo_pub +0ms
  rascal:Publication Initialising publication: /demo_ex +0ms
  rascal:Publication Initialising publication: / +0ms
  rascal:Publication Initialising publication: /demo_q +1ms
  rascal:Subscription Initialising subscription: demo_sub +0ms
  rascal:Subscription Initialising subscription: /demo_q +0ms
  rascal:Subscription Subscribing to queue: demo_q +3ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +5ms
  rascal:Vhost Created confirm channel: efc7cd5a-6d85-494d-acc2-3ce8729f2da8 from connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e +11ms
  rascal:Vhost Created confirm channel: 34123f0c-9412-4cf0-b5c0-104f6f68d42f from connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e +2ms
  rascal:SubscriberSession Opening subscriber session: amq.ctag-sA1tSrml0sZYIhB4RAbZkA on channel: 34123f0c-9412-4cf0-b5c0-104f6f68d42f +0ms
  rascal:Vhost Handling connection error: Unexpected close initially from connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +7s
Error: Unexpected close
    at succeed (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:272:13)
    at onOpenOk (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:254:5)
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:166:32
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:160:12
    at Socket.recv (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:499:12)
    at Object.onceWrapper (events.js:288:20)
    at Socket.emit (events.js:200:13)
    at emitReadable_ (_stream_readable.js:558:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +3ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +7s
  rascal:Subscription Handling channel error: Unexpected close from demo_sub using channel: 34123f0c-9412-4cf0-b5c0-104f6f68d42f +7s
Error: Unexpected close
    at succeed (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:272:13)
    at onOpenOk (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:254:5)
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:166:32
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:160:12
    at Socket.recv (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:499:12)
    at Object.onceWrapper (events.js:288:20)
    at Socket.emit (events.js:200:13)
    at emitReadable_ (_stream_readable.js:558:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9)
  rascal:SubscriberSession Removing channel: 34123f0c-9412-4cf0-b5c0-104f6f68d42f from session +7s
  rascal:Subscription Subscribing to queue: demo_q +3ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +6ms
  rascal:Vhost Will attempt reconnection in in 1964ms +7ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake initially from connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +2s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake
    at Socket.endWhileOpening (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:260:17)
    at Socket.emit (events.js:205:15)
    at endReadableNT (_stream_readable.js:1154:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +2ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +2s
  rascal:Vhost Will attempt reconnection in in 3187ms +2ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: connect ECONNREFUSED 127.0.0.1:5672 initially from connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +3s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: connect ECONNREFUSED 127.0.0.1:5672
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1054:14) {
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 5672
} {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +3s
  rascal:Vhost Will attempt reconnection in in 6258ms +4ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake initially from connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +6s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake
    at Socket.endWhileOpening (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:260:17)
    at Socket.emit (events.js:205:15)
    at endReadableNT (_stream_readable.js:1154:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +6s
  rascal:Vhost Will attempt reconnection in in 10417ms +5ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake initially from connection: fee2a70e-9d8b-44a2-bfba-5d0f64319a6e, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +10s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake
    at Socket.endWhileOpening (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:260:17)
    at Socket.emit (events.js:205:15)
    at endReadableNT (_stream_readable.js:1154:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +10s
  rascal:tasks:createConnection Obtained connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b +14ms
  rascal:tasks:createChannel Creating channel +29s
  rascal:tasks:assertExchanges Asserting exchange: demo_ex +29s
  rascal:tasks:assertQueues Asserting queue: demo_q +29s
  rascal:tasks:applyBindings Binding queue: demo_q to exchange: demo_ex with binding key: a.b.c +29s
  rascal:tasks:closeChannel Closing channel +29s
  rascal:Vhost vhost: / was initialised with connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b +29ms
  rascal:Vhost Created confirm channel: 880943ce-69da-4223-af15-372ac78c7910 from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b +5ms
  rascal:SubscriberSession Opening subscriber session: amq.ctag-T0SP1CYE-9jGtNpuR-umog on channel: 880943ce-69da-4223-af15-372ac78c7910 +22s
  rascal:Vhost Destroying channel: efc7cd5a-6d85-494d-acc2-3ce8729f2da8 +41s
  rascal:Vhost Created confirm channel: 20a90c9a-30d4-4a79-ae85-8acf2a77f8af from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b +8ms
  rascal:Vhost Destroying channel: 20a90c9a-30d4-4a79-ae85-8acf2a77f8af +1m
  rascal:Vhost Created confirm channel: ce17aa1c-21f1-4a36-859e-0ab26fe1b6c5 from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b +9ms
  rascal:Vhost Handling connection error: Unexpected close initially from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +13s
Error: Unexpected close
    at succeed (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:272:13)
    at onOpenOk (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:254:5)
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:166:32
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:160:12
    at Socket.recv (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:499:12)
    at Object.onceWrapper (events.js:288:20)
    at Socket.emit (events.js:200:13)
    at emitReadable_ (_stream_readable.js:558:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +2m
  rascal:Subscription Handling channel error: Unexpected close from demo_sub using channel: 880943ce-69da-4223-af15-372ac78c7910 +2m
Error: Unexpected close
    at succeed (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:272:13)
    at onOpenOk (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:254:5)
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:166:32
    at /Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:160:12
    at Socket.recv (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:499:12)
    at Object.onceWrapper (events.js:288:20)
    at Socket.emit (events.js:200:13)
    at emitReadable_ (_stream_readable.js:558:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9)
  rascal:SubscriberSession Removing channel: 880943ce-69da-4223-af15-372ac78c7910 from session +2m
  rascal:Subscription Subscribing to queue: demo_q +0ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +2ms
  rascal:Vhost Will attempt reconnection in in 1778ms +8ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake initially from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +2s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake
    at Socket.endWhileOpening (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:260:17)
    at Socket.emit (events.js:205:15)
    at endReadableNT (_stream_readable.js:1154:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +2s
  rascal:Vhost Will attempt reconnection in in 2753ms +2ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: connect ECONNREFUSED 127.0.0.1:5672 initially from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +3s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: connect ECONNREFUSED 127.0.0.1:5672
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1054:14) {
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 5672
} {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +3s
  rascal:Vhost Will attempt reconnection in in 5658ms +5ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake initially from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +6s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake
    at Socket.endWhileOpening (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:260:17)
    at Socket.emit (events.js:205:15)
    at endReadableNT (_stream_readable.js:1154:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +6s
  rascal:Vhost Will attempt reconnection in in 12926ms +4ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake initially from connection: c1b8b5b5-4047-4f25-aa63-4d966138ce4b, amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +13s
Error: Failed to connect to: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100. Original message was: Socket closed abruptly during opening handshake
    at Socket.endWhileOpening (/Users/steve/Development/guidesmiths/rascal/node_modules/amqplib/lib/connection.js:260:17)
    at Socket.emit (events.js:205:15)
    at endReadableNT (_stream_readable.js:1154:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:9) {
  vhost: '/',
  connectionUrl: 'amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100'
}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:5672?heartbeat=10&connection_timeout=10000&channelMax=100 +13s
  rascal:tasks:createConnection Obtained connection: 1741b08c-da4e-482d-ab45-244f77ec130a +16ms
  rascal:tasks:createChannel Creating channel +2m
  rascal:tasks:assertExchanges Asserting exchange: demo_ex +2m
  rascal:tasks:assertQueues Asserting queue: demo_q +2m
  rascal:tasks:applyBindings Binding queue: demo_q to exchange: demo_ex with binding key: a.b.c +2m
  rascal:tasks:closeChannel Closing channel +2m
  rascal:Vhost vhost: / was initialised with connection: 1741b08c-da4e-482d-ab45-244f77ec130a +35ms
  rascal:Vhost Created confirm channel: 2f8d914f-93c0-4ee7-a746-5e15d4222b73 from connection: 1741b08c-da4e-482d-ab45-244f77ec130a +7ms
  rascal:SubscriberSession Opening subscriber session: amq.ctag-EDGZuq-5DHmAoALJR2V6Dw on channel: 2f8d914f-93c0-4ee7-a746-5e15d4222b73 +23s
```

I suspect the reason why it appears that there are no retries is because you never actually obtain a connection. Whereas the above debug includes the following...

```
rascal:tasks:createConnection Obtained connection: 1741b08c-da4e-482d-ab45-244f77ec130a +16ms
```

I also notice that your connection string (amqp://guest:***@localhost:31072) doesn't include a connection timeout parameter, e.g. `?connection_timeout=10000`. Maybe there's something going on in your environment which means the node application is never reconnecting, and that the connection attempt is never timing out?

**Comment by cressie176** [comment]
A couple of other things you might consider.

1. Upgrading to rascal@10.2.4 - this version contains fixes/workarounds for a few bugs around connection management (It doesn't look like any of these are causing your problem though, but worth ruling out)

1. Debugging with [WireShark](https://www.wireshark.org/). WireShark includes an amqp filter, so once installed you can easily follow the packets going between your application and RabbitMQ. If I'm write about the connection timeout I expect you will see a "Connection.Open" from the application to the broker, but no corresponding "Connection.Open-OK" reply from the broker to your application.

**Comment by jliu-tera** [comment]
Thanks @cressie176. I will try your suggestions and report back.

**Comment by cressie176** [comment]
Any update @jliu-tera?

**Comment by jliu-tera** [comment]
Sorry @cressie176 I was busy for the last few weeks. I've added timeout and upgraded to Rascal 10.2.6. Still cannot get consistent re-connect.

As you can see from the log below, first the connection was obtained. then I restarted RabbitMQ pod, it tried to re-connect and the connection was obtained again.

Then deleted and restarted RabbitMQ pod again, however no re-connect attempt was found from the log.

`rascal:config:configure Configuring vhost: / +0ms
  rascal:config:configure Configuring exchange: endpoint.out +6ms
  rascal:config:configure Configuring exchange: ditto.out +0ms
  rascal:config:configure Configuring exchange:  +0ms
  rascal:config:configure Configuring queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +1ms
  rascal:config:configure Configuring binding: bind-to-fanout +1ms
  rascal:config:configure Configuring publication: connectorPub +0ms
  rascal:config:configure Configuring subscription: connectorSub +1ms
  rascal:config:configure Configuring publication: connectorPub +0ms
  rascal:config:configure Configuring publication: /endpoint.out +0ms
  rascal:config:configure Configuring publication: /ditto.out +1ms
  rascal:config:configure Configuring publication: / +0ms
  rascal:config:configure Configuring publication: /ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +0ms
  rascal:config:configure Configuring subscription: connectorSub +0ms
  rascal:config:configure Configuring subscription: /ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +0ms
  rascal:config:configure Configuring counter: stub +0ms
  rascal:Broker Initialising broker +0ms
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +0ms
  rascal:tasks:createConnection Obtained connection: 496aa873-7401-4789-ab96-e8b8073831f0 +66ms
  rascal:tasks:createChannel Creating channel +0ms
  rascal:tasks:checkExchanges Checking exchange: endpoint.out +0ms
  rascal:tasks:checkExchanges Checking exchange: ditto.out +11ms
  rascal:tasks:assertQueues Asserting queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +0ms
  rascal:tasks:applyBindings Binding queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 to exchange: ditto.out with binding key: command +0ms
  rascal:tasks:closeChannel Closing channel +0ms
  rascal:Vhost Creating regular channel pool { autostart: false, max: 5, min: 1, evictionRunIntervalMillis: 10000, idleTimeoutMillis: 60000, rejectionDelayMillis: 1000, testOnBorrow: true } +92ms
  rascal:Vhost Creating confirm channel pool { autostart: false, max: 5, min: 1, evictionRunIntervalMillis: 10000, idleTimeoutMillis: 60000, rejectionDelayMillis: 1000, testOnBorrow: true } +3ms
  rascal:Vhost vhost: / was initialised with connection: 496aa873-7401-4789-ab96-e8b8073831f0 +0ms
  rascal:Publication Initialising publication: connectorPub +0ms
  rascal:Publication Initialising publication: /endpoint.out +0ms
  rascal:Publication Initialising publication: /ditto.out +0ms
  rascal:Publication Initialising publication: / +0ms
  rascal:Publication Initialising publication: /ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +0ms
  rascal:Subscription Initialising subscription: connectorSub +0ms
  rascal:Subscription Initialising subscription: /ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +0ms
  rascal:Subscription Subscribing to queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +3ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +6ms
  rascal:Vhost Created confirm channel: 0ce7caf4-8283-4f28-a2e4-0f63952b078a from connection: 496aa873-7401-4789-ab96-e8b8073831f0 +3ms
  rascal:SubscriberSession Opening subscriber session: amq.ctag-lkY-AAAofALWYd_ZOlNtBA on channel: 0ce7caf4-8283-4f28-a2e4-0f63952b078a +0ms
{"message":"Subscribed to Ditto messages","level":"info"}
  rascal:Vhost Handling connection error: write ECONNRESET initially from connection: 496aa873-7401-4789-ab96-e8b8073831f0, amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +23s
{"message":"Broker error: Error: write ECONNRESET; vhost: /; connectionUrl: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000","level":"error"}
  rascal:Vhost Initialising vhost: / +2ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +23s
  rascal:Subscription Handling channel error: write ECONNRESET from connectorSub using channel: 0ce7caf4-8283-4f28-a2e4-0f63952b078a +23s
{"message":"Subscriber error: Error: write ECONNRESET","level":"error"}
  rascal:SubscriberSession Removing channel: 0ce7caf4-8283-4f28-a2e4-0f63952b078a from session +23s
  rascal:Subscription Subscribing to queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +13ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +18ms
  rascal:Vhost Will attempt reconnection in in 1282ms +1s
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000. Original message was: connect ECONNREFUSED 127.0.0.1:31072 initially from connection: 496aa873-7401-4789-ab96-e8b8073831f0, amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +1s
{"message":"Broker error: Error: Failed to connect to: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000. Original message was: connect ECONNREFUSED 127.0.0.1:31072; vhost: /; connectionUrl: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000","level":"error"}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +2s
  rascal:Vhost Will attempt reconnection in in 2701ms +1ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000. Original message was: connect ECONNREFUSED 127.0.0.1:31072 initially from connection: 496aa873-7401-4789-ab96-e8b8073831f0, amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +3s
{"message":"Broker error: Error: Failed to connect to: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000. Original message was: connect ECONNREFUSED 127.0.0.1:31072; vhost: /; connectionUrl: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000","level":"error"}
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +3s
  rascal:Vhost Will attempt reconnection in in 5250ms +13ms
  rascal:Vhost Handling connection error: Failed to connect to: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000. Original message was: connect ECONNREFUSED 127.0.0.1:31072 initially from connection: 496aa873-7401-4789-ab96-e8b8073831f0, amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +5s
{"message":"Broker error: Error: Failed to connect to: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000. Original message was: connect ECONNREFUSED 127.0.0.1:31072; vhost: /; connectionUrl: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000","level":"error"}
  rascal:Vhost Initialising vhost: / +1ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +5s
  rascal:tasks:createConnection Obtained connection: 9c0e05fb-1421-443c-998a-5f9207739f84 +6ms
  rascal:tasks:createChannel Creating channel +33s
  rascal:tasks:checkExchanges Checking exchange: endpoint.out +33s
  rascal:tasks:checkExchanges Checking exchange: ditto.out +0ms
  rascal:tasks:assertQueues Asserting queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +33s
  rascal:tasks:applyBindings Binding queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 to exchange: ditto.out with binding key: command +33s
  rascal:tasks:closeChannel Closing channel +33s
  rascal:Vhost vhost: / was initialised with connection: 9c0e05fb-1421-443c-998a-5f9207739f84 +13ms
  rascal:Vhost Created confirm channel: 827a639c-ece6-4e25-a0f3-af91a8427657 from connection: 9c0e05fb-1421-443c-998a-5f9207739f84 +2ms
  rascal:SubscriberSession Opening subscriber session: amq.ctag-z1hhVtE_V4A0Js99OSgnRA on channel: 827a639c-ece6-4e25-a0f3-af91a8427657 +10s
  rascal:Vhost Handling connection error: write ECONNRESET initially from connection: 9c0e05fb-1421-443c-998a-5f9207739f84, amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +42s
{"message":"Broker error: Error: write ECONNRESET; vhost: /; connectionUrl: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000","level":"error"}
  rascal:Vhost Initialising vhost: / +0ms
  rascal:tasks:createConnection Connecting to broker using url: amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000 +42s
  rascal:Subscription Handling channel error: write ECONNRESET from connectorSub using channel: 827a639c-ece6-4e25-a0f3-af91a8427657 +52s
{"message":"Subscriber error: Error: write ECONNRESET","level":"error"}
  rascal:SubscriberSession Removing channel: 827a639c-ece6-4e25-a0f3-af91a8427657 from session +42s
  rascal:Subscription Subscribing to queue: ec-in-queue-a6635146-995e-4bbb-85b7-ef1fae58b421 +1ms
  rascal:Vhost Requested channel. Outstanding channel requests: 1 +2ms`

**Comment by jliu-tera** [comment]
Thanks @cressie176 for reopening the ticket. Below is the latest broker config I'm using:
```
{
  vhosts: {
    [config.RABBITMQ_VHOST]: {
      connection: {
        url: `${config.RABBITMQ_CONNECTION_STRING}`
      },
      exchanges: {
        [config.RABBITMQ_EXCHANGE_ENDPOINT_OUT]: {
          type: "topic",
          assert: false,
          check: true
        },
        [config.RABBITMQ_EXCHANGE_DITTO_OUT]: {
          type: "fanout",
          assert: false,
          check: true
        },
      },
      queues: {
        [`${inQueueName}`]: {
          assert: true,
          "options": {
            "durable": true,
            "exclusive": true
          }
        },
      },
      bindings: {
        "bind-to-fanout": {
          source: `${config.RABBITMQ_EXCHANGE_DITTO_OUT}`,
          destination: `${inQueueName}`,
          destinationType: "queue",
          bindingKey: "command",
        }
      },
      publications: {
        connectorPub: {
          exchange: `${config.RABBITMQ_EXCHANGE_ENDPOINT_OUT}`,
          routingKey: `${config.RABBITMQ_PUBLISH_ROUTING_KEY}`
        }
      },
      subscriptions: {
        connectorSub: {
          queue: `${inQueueName}`,
          contentType: "application/json",
          options: {
            // ack messages
            noAck: true
          }
        }
      }
    }
  }
}
```

**Comment by jliu-tera** [comment]
I noticed that `Subscription Subscribing to queue` was called even after connection was reset. This seems unreasonable to me. So I tried disabling auto-delete (or exclusive) queue, but it doesn't make any difference.

**Comment by willthrom** [comment]
If I remember correctly,  the connection pool is an independence entity. We checked in the Subscribing if the connection is there....

Regarding the retries, @jliu-tera  have you tried to play with the config?

https://github.com/guidesmiths/rascal#connection
There are two possible policies, you might be using the default which might get you up to 60s waiting for retrying...

**Comment by willthrom** [comment]
I took a look to the Rascal Library and  notice the timer (backoff) for the Connection is never call to be reset.
Something is happening under subscription timer.... This might be the issue. After a re-connection is never reset to the lower value, so in this case, it will always wait as much as left from the last reconnection....

(A wild guess thought, JS is far from my language of expertise...)

**Comment by jliu-tera** [comment]
@willthrom yes I've played with retry configs with both linear and exponential. Most times it doesn't even do any retry
```
      connection: {
        url: `${config.RABBITMQ_CONNECTION_STRING}`,
        "retry": {
          "min": 1000,
          "max": 5000,
          "strategy": "linear"
        }
      },
```

**Comment by jliu-tera** [comment]
I have some good news. I think it is related to how broker is created.  I compared my code to Rascal examples and tried the following methods:

```
    // Method 1
    // Broker is created as in https://github.com/guidesmiths/rascal/blob/master/examples/promises/index.js
    // Re-connect always works
    const Rascal = require('rascal');
    broker = await Rascal.BrokerAsPromised.create(Rascal.withDefaultConfig(brokerConfig));
```
```
    // Method 2
    // Broker is created as in https://www.npmjs.com/package/rascal#asyncawait
    // Re-connect doesn't always work
    const Broker = require('rascal').BrokerAsPromised;
    broker = await Broker.create(brokerConfig);
```
Maybe with method 2 some of the default config are not set? For reference, I did try connection's `retry` config with method 2 but it didn't work.

**Comment by cressie176** [comment]
Hi @jliu-tera,

Good find. Once you've instantiated the broker you can output the final config using `console.log(JSON.stringify(broker.config, null, 2))`. Are there any obvious differences?

**Comment by jliu-tera** [comment]
broker.config:
method 1:
```
{
  "vhosts": {
    "/": {
      "exchanges": {
        "endpoint.out": {
          "type": "topic",
          "assert": false,
          "check": true,
          "name": "endpoint.out",
          "fullyQualifiedName": "endpoint.out",
          "options": {}
        },
        "ditto.out": {
          "type": "fanout",
          "assert": false,
          "check": true,
          "name": "ditto.out",
          "fullyQualifiedName": "ditto.out",
          "options": {}
        },
        "": {
          "name": "",
          "fullyQualifiedName": "",
          "assert": true,
          "type": "topic",
          "options": {}
        }
      },
      "queues": {
        "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01": {
          "assert": true,
          "options": {
            "durable": true,
            "arguments": {}
          },
          "name": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
          "fullyQualifiedName": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01"
        }
      },
      "bindings": {
        "bind-to-fanout": {
          "bindingKey": "command",
          "source": "ditto.out",
          "destination": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
          "destinationType": "queue",
          "name": "bind-to-fanout",
          "options": {}
        }
      },
      "name": "/",
      "connectionStrategy": "random",
      "publicationChannelPools": {
        "regularPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        },
        "confirmPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        }
      },
      "defaults": {
        "connection": {
          "options": {
            "heartbeat": 10,
            "connection_timeout": 10000,
            "channelMax": 100
          },
          "socketOptions": {
            "timeout": 10000
          },
          "management": {
            "options": {
              "timeout": 1000
            },
            "slashes": true,
            "protocol": "http",
            "port": 15672
          },
          "slashes": true,
          "protocol": "amqp",
          "hostname": "localhost",
          "user": "guest",
          "password": "guest",
          "port": "5672",
          "retry": {
            "min": 1000,
            "max": 60000,
            "factor": 2,
            "strategy": "exponential"
          }
        },
        "publicationChannelPools": {
          "regularPool": {
            "autostart": false,
            "max": 5,
            "min": 1,
            "evictionRunIntervalMillis": 10000,
            "idleTimeoutMillis": 60000,
            "rejectionDelayMillis": 1000,
            "testOnBorrow": true
          },
          "confirmPool": {
            "autostart": false,
            "max": 5,
            "min": 1,
            "evictionRunIntervalMillis": 10000,
            "idleTimeoutMillis": 60000,
            "rejectionDelayMillis": 1000,
            "testOnBorrow": true
          }
        },
        "connectionStrategy": "random",
        "exchanges": {
          "assert": true,
          "type": "topic",
          "options": {}
        },
        "queues": {
          "assert": true,
          "options": {
            "arguments": {}
          }
        },
        "bindings": {
          "destinationType": "queue",
          "bindingKey": "#",
          "options": {}
        }
      },
      "connections": [
        {
          "url": "amqp://guest:guest@localhost:31072?heartbeat=10&connection_timeout=5000",
          "retry": {
            "min": 1000,
            "max": 5000,
            "strategy": "linear",
            "factor": 2
          },
          "options": {
            "heartbeat": 10,
            "connection_timeout": 10000,
            "channelMax": 100
          },
          "socketOptions": {
            "timeout": 10000
          },
          "management": {
            "options": {
              "timeout": 1000
            },
            "slashes": true,
            "protocol": "http",
            "port": 15672,
            "hostname": "localhost",
            "auth": "guest:guest",
            "url": "http://guest:guest@localhost:15672",
            "loggableUrl": "http://guest:***@localhost:15672"
          },
          "slashes": true,
          "protocol": "amqp",
          "hostname": "localhost",
          "user": "guest",
          "password": "guest",
          "port": "5672",
          "vhost": "/",
          "auth": "guest:guest",
          "pathname": "",
          "query": {
            "heartbeat": 10,
            "connection_timeout": 10000,
            "channelMax": 100
          },
          "loggableUrl": "amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000",
          "index": 0.2635723585745724
        }
      ]
    }
  },
  "defaults": {
    "vhosts": {
      "connection": {
        "options": {
          "heartbeat": 10,
          "connection_timeout": 10000,
          "channelMax": 100
        },
        "socketOptions": {
          "timeout": 10000
        },
        "management": {
          "options": {
            "timeout": 1000
          },
          "slashes": true,
          "protocol": "http",
          "port": 15672
        },
        "slashes": true,
        "protocol": "amqp",
        "hostname": "localhost",
        "user": "guest",
        "password": "guest",
        "port": "5672",
        "retry": {
          "min": 1000,
          "max": 60000,
          "factor": 2,
          "strategy": "exponential"
        }
      },
      "publicationChannelPools": {
        "regularPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        },
        "confirmPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        }
      },
      "connectionStrategy": "random",
      "exchanges": {
        "assert": true,
        "type": "topic",
        "options": {}
      },
      "queues": {
        "assert": true,
        "options": {
          "arguments": {}
        }
      },
      "bindings": {
        "destinationType": "queue",
        "bindingKey": "#",
        "options": {}
      }
    },
    "publications": {
      "vhost": "/",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      }
    },
    "subscriptions": {
      "vhost": "/",
      "prefetch": 10,
      "retry": {
        "min": 1000,
        "max": 60000,
        "factor": 2,
        "strategy": "exponential"
      },
      "redeliveries": {
        "limit": 100,
        "timeout": 1000,
        "counter": "stub"
      },
      "deferCloseChannel": 10000,
      "options": {}
    },
    "redeliveries": {
      "counters": {
        "stub": {},
        "inMemory": {
          "size": 1000
        }
      }
    },
    "shovels": {}
  },
  "publications": {
    "connectorPub": {
      "exchange": "endpoint.out",
      "routingKey": "connector",
      "vhost": "/",
      "name": "connectorPub",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "endpoint.out"
    },
    "/endpoint.out": {
      "exchange": "endpoint.out",
      "vhost": "/",
      "autoCreated": true,
      "name": "/endpoint.out",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "endpoint.out"
    },
    "/ditto.out": {
      "exchange": "ditto.out",
      "vhost": "/",
      "autoCreated": true,
      "name": "/ditto.out",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "ditto.out"
    },
    "/": {
      "exchange": "",
      "vhost": "/",
      "autoCreated": true,
      "name": "/",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": ""
    },
    "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01": {
      "queue": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
      "vhost": "/",
      "autoCreated": true,
      "name": "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01"
    }
  },
  "subscriptions": {
    "connectorSub": {
      "queue": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
      "contentType": "application/json",
      "options": {
        "noAck": true
      },
      "vhost": "/",
      "name": "connectorSub",
      "prefetch": 10,
      "retry": {
        "min": 1000,
        "max": 60000,
        "factor": 2,
        "strategy": "exponential"
      },
      "redeliveries": {
        "limit": 100,
        "timeout": 1000,
        "counter": "stub"
      },
      "deferCloseChannel": 10000,
      "source": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
      "encryption": {}
    },
    "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01": {
      "queue": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
      "vhost": "/",
      "autoCreated": true,
      "name": "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
      "prefetch": 10,
      "retry": {
        "min": 1000,
        "max": 60000,
        "factor": 2,
        "strategy": "exponential"
      },
      "redeliveries": {
        "limit": 100,
        "timeout": 1000,
        "counter": "stub"
      },
      "deferCloseChannel": 10000,
      "options": {},
      "source": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
      "encryption": {}
    }
  },
  "redeliveries": {
    "counters": {
      "stub": {
        "name": "stub",
        "type": "stub"
      }
    }
  }
}
```

method 2:
```
{
  "vhosts": {
    "/": {
      "exchanges": {
        "endpoint.out": {
          "type": "topic",
          "assert": false,
          "check": true,
          "name": "endpoint.out",
          "fullyQualifiedName": "endpoint.out",
          "options": {}
        },
        "ditto.out": {
          "type": "fanout",
          "assert": false,
          "check": true,
          "name": "ditto.out",
          "fullyQualifiedName": "ditto.out",
          "options": {}
        },
        "": {
          "name": "",
          "fullyQualifiedName": "",
          "assert": true,
          "type": "topic",
          "options": {}
        }
      },
      "queues": {
        "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a": {
          "assert": true,
          "options": {
            "durable": true,
            "arguments": {}
          },
          "name": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
          "fullyQualifiedName": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a"
        }
      },
      "bindings": {
        "bind-to-fanout": {
          "bindingKey": "command",
          "source": "ditto.out",
          "destination": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
          "destinationType": "queue",
          "name": "bind-to-fanout",
          "options": {}
        }
      },
      "name": "/",
      "connectionStrategy": "random",
      "publicationChannelPools": {
        "regularPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        },
        "confirmPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        }
      },
      "defaults": {
        "publicationChannelPools": {
          "regularPool": {
            "autostart": false,
            "max": 5,
            "min": 1,
            "evictionRunIntervalMillis": 10000,
            "idleTimeoutMillis": 60000,
            "rejectionDelayMillis": 1000,
            "testOnBorrow": true
          },
          "confirmPool": {
            "autostart": false,
            "max": 5,
            "min": 1,
            "evictionRunIntervalMillis": 10000,
            "idleTimeoutMillis": 60000,
            "rejectionDelayMillis": 1000,
            "testOnBorrow": true
          }
        },
        "connectionStrategy": "random",
        "connection": {
          "slashes": true,
          "protocol": "amqp",
          "hostname": "localhost",
          "user": "guest",
          "password": "guest",
          "port": "5672",
          "options": {},
          "retry": {
            "min": 1000,
            "max": 60000,
            "factor": 2,
            "strategy": "exponential"
          },
          "management": {
            "slashes": true,
            "protocol": "http",
            "port": 15672,
            "options": {}
          }
        },
        "exchanges": {
          "assert": true,
          "type": "topic",
          "options": {}
        },
        "queues": {
          "assert": true,
          "options": {
            "arguments": {}
          }
        },
        "bindings": {
          "destinationType": "queue",
          "bindingKey": "#",
          "options": {}
        }
      },
      "connections": [
        {
          "url": "amqp://guest:guest@localhost:31072?heartbeat=10&connection_timeout=5000",
          "retry": {
            "min": 1000,
            "max": 5000,
            "strategy": "linear",
            "factor": 2
          },
          "slashes": true,
          "protocol": "amqp",
          "hostname": "localhost",
          "user": "guest",
          "password": "guest",
          "port": "5672",
          "options": {},
          "management": {
            "slashes": true,
            "protocol": "http",
            "port": 15672,
            "options": {},
            "hostname": "localhost",
            "auth": "guest:guest",
            "url": "http://guest:guest@localhost:15672",
            "loggableUrl": "http://guest:***@localhost:15672"
          },
          "vhost": "/",
          "auth": "guest:guest",
          "pathname": "",
          "query": {},
          "loggableUrl": "amqp://guest:***@localhost:31072?heartbeat=10&connection_timeout=5000",
          "index": 0.8311549829934703
        }
      ]
    }
  },
  "defaults": {
    "vhosts": {
      "publicationChannelPools": {
        "regularPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        },
        "confirmPool": {
          "autostart": false,
          "max": 5,
          "min": 1,
          "evictionRunIntervalMillis": 10000,
          "idleTimeoutMillis": 60000,
          "rejectionDelayMillis": 1000,
          "testOnBorrow": true
        }
      },
      "connectionStrategy": "random",
      "connection": {
        "slashes": true,
        "protocol": "amqp",
        "hostname": "localhost",
        "user": "guest",
        "password": "guest",
        "port": "5672",
        "options": {},
        "retry": {
          "min": 1000,
          "max": 60000,
          "factor": 2,
          "strategy": "exponential"
        },
        "management": {
          "slashes": true,
          "protocol": "http",
          "port": 15672,
          "options": {}
        }
      },
      "exchanges": {
        "assert": true,
        "type": "topic",
        "options": {}
      },
      "queues": {
        "assert": true,
        "options": {
          "arguments": {}
        }
      },
      "bindings": {
        "destinationType": "queue",
        "bindingKey": "#",
        "options": {}
      }
    },
    "publications": {
      "vhost": "/",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      }
    },
    "subscriptions": {
      "vhost": "/",
      "prefetch": 10,
      "retry": {
        "min": 1000,
        "max": 60000,
        "factor": 2,
        "strategy": "exponential"
      },
      "redeliveries": {
        "limit": 100,
        "timeout": 1000,
        "counter": "stub"
      },
      "deferCloseChannel": 10000,
      "options": {}
    },
    "redeliveries": {
      "counters": {
        "stub": {},
        "inMemory": {
          "size": 1000
        }
      }
    },
    "shovels": {}
  },
  "publications": {
    "connectorPub": {
      "exchange": "endpoint.out",
      "routingKey": "connector",
      "vhost": "/",
      "name": "connectorPub",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "endpoint.out"
    },
    "/endpoint.out": {
      "exchange": "endpoint.out",
      "vhost": "/",
      "autoCreated": true,
      "name": "/endpoint.out",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "endpoint.out"
    },
    "/ditto.out": {
      "exchange": "ditto.out",
      "vhost": "/",
      "autoCreated": true,
      "name": "/ditto.out",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "ditto.out"
    },
    "/": {
      "exchange": "",
      "vhost": "/",
      "autoCreated": true,
      "name": "/",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": ""
    },
    "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a": {
      "queue": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
      "vhost": "/",
      "autoCreated": true,
      "name": "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
      "confirm": true,
      "timeout": 10000,
      "options": {
        "persistent": true,
        "mandatory": true
      },
      "destination": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a"
    }
  },
  "subscriptions": {
    "connectorSub": {
      "queue": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
      "contentType": "application/json",
      "options": {
        "noAck": true
      },
      "vhost": "/",
      "name": "connectorSub",
      "prefetch": 10,
      "retry": {
        "min": 1000,
        "max": 60000,
        "factor": 2,
        "strategy": "exponential"
      },
      "redeliveries": {
        "limit": 100,
        "timeout": 1000,
        "counter": "stub"
      },
      "deferCloseChannel": 10000,
      "source": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
      "encryption": {}
    },
    "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a": {
      "queue": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
      "vhost": "/",
      "autoCreated": true,
      "name": "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
      "prefetch": 10,
      "retry": {
        "min": 1000,
        "max": 60000,
        "factor": 2,
        "strategy": "exponential"
      },
      "redeliveries": {
        "limit": 100,
        "timeout": 1000,
        "counter": "stub"
      },
      "deferCloseChannel": 10000,
      "options": {},
      "source": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
      "encryption": {}
    }
  },
  "redeliveries": {
    "counters": {
      "stub": {
        "name": "stub",
        "type": "stub"
      }
    }
  }
}
```

**Comment by jliu-tera** [comment]
And the diff:
```
30c30
<         "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01": {
---
>         "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a": {
36,37c36,37
<           "name": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
<           "fullyQualifiedName": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01"
---
>           "name": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
>           "fullyQualifiedName": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a"
44c44
<           "destination": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>           "destination": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
73,102d72
<         "connection": {
<           "options": {
<             "heartbeat": 10,
<             "connection_timeout": 10000,
<             "channelMax": 100
<           },
<           "socketOptions": {
<             "timeout": 10000
<           },
<           "management": {
<             "options": {
<               "timeout": 1000
<             },
<             "slashes": true,
<             "protocol": "http",
<             "port": 15672
<           },
<           "slashes": true,
<           "protocol": "amqp",
<           "hostname": "localhost",
<           "user": "guest",
<           "password": "guest",
<           "port": "5672",
<           "retry": {
<             "min": 1000,
<             "max": 60000,
<             "factor": 2,
<             "strategy": "exponential"
<           }
<         },
123a94,114
>         "connection": {
>           "slashes": true,
>           "protocol": "amqp",
>           "hostname": "localhost",
>           "user": "guest",
>           "password": "guest",
>           "port": "5672",
>           "options": {},
>           "retry": {
>             "min": 1000,
>             "max": 60000,
>             "factor": 2,
>             "strategy": "exponential"
>           },
>           "management": {
>             "slashes": true,
>             "protocol": "http",
>             "port": 15672,
>             "options": {}
>           }
>         },
150,157c141,147
<           "options": {
<             "heartbeat": 10,
<             "connection_timeout": 10000,
<             "channelMax": 100
<           },
<           "socketOptions": {
<             "timeout": 10000
<           },
---
>           "slashes": true,
>           "protocol": "amqp",
>           "hostname": "localhost",
>           "user": "guest",
>           "password": "guest",
>           "port": "5672",
>           "options": {},
159,161d148
<             "options": {
<               "timeout": 1000
<             },
164a152
>             "options": {},
170,175d157
<           "slashes": true,
<           "protocol": "amqp",
<           "hostname": "localhost",
<           "user": "guest",
<           "password": "guest",
<           "port": "5672",
179,183c161
<           "query": {
<             "heartbeat": 10,
<             "connection_timeout": 10000,
<             "channelMax": 100
<           },
---
>           "query": {},
185c163
<           "index": 0.2635723585745724
---
>           "index": 0.8311549829934703
192,221d169
<       "connection": {
<         "options": {
<           "heartbeat": 10,
<           "connection_timeout": 10000,
<           "channelMax": 100
<         },
<         "socketOptions": {
<           "timeout": 10000
<         },
<         "management": {
<           "options": {
<             "timeout": 1000
<           },
<           "slashes": true,
<           "protocol": "http",
<           "port": 15672
<         },
<         "slashes": true,
<         "protocol": "amqp",
<         "hostname": "localhost",
<         "user": "guest",
<         "password": "guest",
<         "port": "5672",
<         "retry": {
<           "min": 1000,
<           "max": 60000,
<           "factor": 2,
<           "strategy": "exponential"
<         }
<       },
242a191,211
>       "connection": {
>         "slashes": true,
>         "protocol": "amqp",
>         "hostname": "localhost",
>         "user": "guest",
>         "password": "guest",
>         "port": "5672",
>         "options": {},
>         "retry": {
>           "min": 1000,
>           "max": 60000,
>           "factor": 2,
>           "strategy": "exponential"
>         },
>         "management": {
>           "slashes": true,
>           "protocol": "http",
>           "port": 15672,
>           "options": {}
>         }
>       },
349,350c318,319
<     "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01": {
<       "queue": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>     "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a": {
>       "queue": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
353c322
<       "name": "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>       "name": "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
360c329
<       "destination": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01"
---
>       "destination": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a"
365c334
<       "queue": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>       "queue": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
385c354
<       "source": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>       "source": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
388,389c357,358
<     "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01": {
<       "queue": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>     "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a": {
>       "queue": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
392c361
<       "name": "/ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>       "name": "/ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
407c376
<       "source": "ec-in-queue-38edc47c-fec2-42ba-8560-8166b0388c01",
---
>       "source": "ec-in-queue-fe818904-5679-4419-b00a-1758ebb18c2a",
```

**Comment by jliu-tera** [comment]
It's hard for me to figure out the difference. They seems just to be in different orders.

**Comment by cressie176** [comment]
If you use a tool like [jsondiff](http://www.jsondiff.com/) you'll see the differences a bit more easily. The big ones are around the connection and socket options. The following are missing when you don't use `withDefaultConfig`

```json
{
  "connection": {
    "options": {
      "channelMax": 100,
      "connection_timeout": 10000,
      "heartbeat": 10
    },
    "socketOptions": {
      "timeout": 10000
    }
  }
}
```

Without the connection / socket timeout there's nothing to tell node to give up if it can't reconnect. Suspect this is related.

**Comment by jliu-tera** [comment]
Yes that should be the problem. I added connection/socket timeout in connection config, it re-connects with method 2.

**Comment by jliu-tera** [comment]
@cressie176, I have an issue related to this one, `broker.on('vhost_initialized')` or `broker.on('ready')` are never called after broker is connected.

The above 2 callbacks are used to detect whether re-connect is successful. Is something wrong with my config, or is this a bug?

Below is the code snippet
```
   broker = await Rascal.BrokerAsPromised.create(Rascal.withDefaultConfig(brokerConfig));
    brokerReady = true;

    broker.on('busy', ({vhost}) => {
      if (vhost === config.RABBITMQ_VHOST) {
        brokerReady = false;
      }
    });
    broker.on('ready', ({vhost}) => {
      if (vhost === config.RABBITMQ_VHOST) {
        brokerReady = true;
      }
    });
    broker.on('blocked', (reason, {vhost, connectionUrl}) => {
      if (vhost === config.RABBITMQ_VHOST) {
        brokerReady = false;
      }
    });
    broker.on('unblocked', ({vhost, connectionUrl}) => {
      if (vhost === config.RABBITMQ_VHOST) {
        brokerReady = true;
      }
    });

    broker.on('vhost_initialized', ({vhost, connectionUrl}) => {
      if (vhost === config.RABBITMQ_VHOST) {
        brokerReady = true;
      }
    });

    broker.on('error', (error, {vhost, connectionUrl}) => {
      if (vhost === config.RABBITMQ_VHOST) {
        brokerReady = false;
      }
    });
  } catch (error) {
    logger.error(`Rascal broker config error: ${error.message}`);
  }
```

**Comment by jliu-tera** [comment]
on('error') always gets called when RabbitMQ is killed.

**Comment by jliu-tera** [comment]
Found the problem: `vhost_initialized` should be `vhost_initialised`. My bad, or spelling checker's fault.

It all works as I expected now. Please close the ticket. Thanks a lot for the great library and your time on this issue.

**Comment by cressie176** [comment]
Glad you found it. I've done a very similar things with the HTTP "Authorization" header.
